include/oneapi/dpl/pstl/iterator_impl.h

```diff
@@ -410,19 +414,36 @@ class transform_iterator
     typedef typename ::std::iterator_traits<_Iter>::pointer pointer;
     typedef typename ::std::iterator_traits<_Iter>::iterator_category iterator_category;
 
-    transform_iterator(_Iter __it = _Iter(), _UnaryFunc __unary_func = _UnaryFunc())
-        : __my_it_(__it), __my_unary_func_(__unary_func)
+    //default constructor will only be present if both the unary functor and iterator are default constructible
+    transform_iterator() = default;
+
+    //only enable this constructor if the unary functor is default constructible
+    template <typename _UnaryFuncLocal = _UnaryFunc,
+              ::std::enable_if_t<::std::is_default_constructible_v<_UnaryFuncLocal>, int> = 0>
+    transform_iterator(_Iter __it) : __my_it_(std::move(__it))
     {
     }
-    transform_iterator(const transform_iterator& __input) = default;
+
+    transform_iterator(_Iter __it, _UnaryFunc __unary_func)
+        : __my_it_(std::move(__it)), __my_unary_func_(std::move(__unary_func))
+    {
+    }
+
+    transform_iterator(const transform_iterator&) = default;
     transform_iterator&
     operator=(const transform_iterator& __input)
     {
-        //TODO: Investigate making transform_iterator trivially_copyable. This custom copy assignment operator prevents
-        // transform_iterator, and therefore permutation_iterator from being trivially_copyable.  Not being trivially
-        // copyable makes their device_copyable trait deprecated in SYCL2020. However, defaulting this function implies
-        // an extra requirement that __my_unary_func_ implements a copy assignment operator.
         __my_it_ = __input.__my_it_;
+
+        // If copy assignment is available, copy the functor, otherwise skip it.
+        // For non-copy assignable functors, this copy assignment operator departs from the sycl 2020 specification
+        // requirement of device copyable types for copy assignment to be the same as a bitwise copy of the object.
+        // TODO: Explore (ABI breaking) change to use std::optional or similar and using copy constructor to implement
+        //       copy assignment to better comply with SYCL 2020 specification.
+        if constexpr (std::is_copy_assignable_v<_UnaryFunc>)
```

Comment:
In the opposite case, should we initialize `__my_unary_func_`  ?

Reply:
Maybe I'm missing something, but I'm not sure how we would do that.  We are in the copy assignment operator, and it is not copy assignable.  
We can't construct it without the `std::optional` -like mechanism, which we have decided not to pursue right now.

Do you have another method in mind?

Reply:
I think that we can try to do something like this:
```C++
        if constexpr (std::is_copy_assignable_v<_UnaryFunc>)
        {
            __my_unary_func_ = __input.__my_unary_func_;
        }
        else
        {
            __my_unary_func_ = _UnaryFunc{};
        }
```

Reply:
Isn't that using the copy assignment?

https://godbolt.org/z/nGTGe4rqE

Reply:
Technically, we could do something like in the following link (check default constructibility and move assignability, and use them if avail).  However, this seems like it only serves a very unlikely scenario with an unexpected result (default construction rather than copy within the copy assignment operator).  I think that we are better off just doing nothing when we are missing copy assignment.

https://godbolt.org/z/vbzhGK9T1

Reply:
@dan, of course, it's up to you, Bit I think that implementation of `operator=` where something assigned and other part sometimes save their state looks like a little bit incorrect because you allow through constructor to have state for this "unassigned" part.

Reply:
@rarutyun and I discussed this in some detail offline, and I perhaps didn't fully describe that here in the PR comments.  I will try to state some of that here.

Some other options:
1) leave as is  (worse for copy assignable functors with regard to device copyable sycl spec)
2) delete copy assignment when functor has no copy assignment (potential for user code breaking with c++17 lambdas, see #1437)
3) std::optional (ABI breaking)

We are not adding this operator as new functionality in this PR, its previous state was to *always* skip copying of the functor. In fact, we are shrinking the number of cases which have this skipped copy of the functor (prior to the PR, all functors skip this copy).  The PR remedies many cases where copy assignment is available, and makes them in alignment with the SYCL specification for device copyable.

This change also does not break ABI, or break any existing user code (the one exception could be users which depend upon the functor *not* being copied even when it is copy assignable, but that seems far-fetched).

We are not preventing future changes as indicated in the "TODO" to use something like a `std::optional` and using copy construction to implement copy assignment where it is possible.  However, it seems that such a change would require more thought and is ABI breaking.

I agree with your point that this unbalanced copy assignment is not ideal, but it seems the best option at the current time with the context we have.

Reply:
ok. Thank you for this clarification.